Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIActionCocoa.mm

```diff
@@ -0,0 +1,305 @@
+        return;
+    }
+
+    ASSERT(action);
+
+    action->setDisplayLabel(title);
+
+    completionHandler(std::nullopt);
+}
+
+void WebExtensionContext::actionSetIcon(std::optional<WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebExtensionTabIdentifier> tabIdentifier, const String& iconDictionaryJSON, CompletionHandler<void(std::optional<String>)>&& completionHandler)
+{
+    static NSString * const apiName = @"action.setIcon()";
+
+    std::optional<String> error;
+    auto action = getOrCreateActionWithIdentifiers(windowIdentifier, tabIdentifier, *this, apiName, error);
+    if (error) {
+        completionHandler(error);
+        return;
+    }
+
+    ASSERT(action);
+
+    action->setIconsDictionary(parseJSON(iconDictionaryJSON));
+
+    completionHandler(std::nullopt);
+}
+
+void WebExtensionContext::actionGetPopup(std::optional<WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebExtensionTabIdentifier> tabIdentifier, CompletionHandler<void(std::optional<String>, std::optional<String>)>&& completionHandler)
+{
+    static NSString * const apiName = @"action.getPopup()";
+
+    std::optional<String> error;
+    auto action = getActionWithIdentifiers(windowIdentifier, tabIdentifier, *this, apiName, error);
+    if (error) {
+        completionHandler(std::nullopt, error);
+        return;
+    }
+
+    ASSERT(action);
+
+    completionHandler(action->popupPath(), std::nullopt);
+}
+
+void WebExtensionContext::actionSetPopup(std::optional<WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebExtensionTabIdentifier> tabIdentifier, const String& popupPath, CompletionHandler<void(std::optional<String>)>&& completionHandler)
+{
+    static NSString * const apiName = @"action.setPopup()";
+
+    std::optional<String> error;
+    auto action = getOrCreateActionWithIdentifiers(windowIdentifier, tabIdentifier, *this, apiName, error);
+    if (error) {
+        completionHandler(error);
+        return;
+    }
+
+    ASSERT(action);
+
+    action->setPopupPath(popupPath);
+
+    completionHandler(std::nullopt);
+}
+
+void WebExtensionContext::actionOpenPopup(WebPageProxyIdentifier identifier, std::optional<WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebExtensionTabIdentifier> tabIdentifier, CompletionHandler<void(std::optional<String>)>&& completionHandler)
+{
+    static NSString * const apiName = @"action.openPopup()";
+
+    if (!defaultAction().canProgramaticlyPresentPopup()) {
+        completionHandler(toErrorString(apiName, nil, @"it is not implemented"));
+        return;
+    }
+
+    if (windowIdentifier) {
+        auto window = getWindow(windowIdentifier.value());
+        if (!window) {
+            completionHandler(toErrorString(apiName, nil, @"window not found"));
+            return;
+        }
+
+        if (auto activeTab = window->activeTab()) {
+            if (getAction(activeTab.get())->hasPopup())
+                performAction(activeTab.get(), UserTriggered::No);
```

Comment:
Do we want to check at the top of this method that we _have_ a popup? Or could this also perform a click action if the extension has specified that.